### PR TITLE
async all the things

### PR DESCRIPTION
**Pull request by lewis6991**

### CODE REVIEW

File: lua/nvim-treesitter/install.lua [hunk 1]
```diff
@@ -67,11 +68,15 @@ local function get_parser_install_info(lang, validate)
   return install_info
 end
 
+local function get_package_path(...)
```
Comment (clason) [inline lua/nvim-treesitter/install.lua:70]: this is used in `./scripts/write-lockfile.lua`, btw.

(but could just as well be inlined there, of course)

**Comment by clason** [comment]
What do you think of hard-coding the compiler + arguments per platform, and exposing a variable that allows providing a function that takes the necessary arguments (files, repo location, C++ scanner etc) and returns a custom compile invocation?

(If we can somehow compile the `norg` parser without shenanigans, we can remove half of the infracstructure here...)

File: lua/nvim-treesitter/install.lua [hunk 2]
```diff
@@ -457,16 +560,25 @@ end
...
 ---@field skip table
 
+local unlink = a.wrap(uv.fs_unlink, 2)
+local symlink = a.wrap(uv.fs_symlink, 4)
+
 -- Install a parser
 ---@param languages? string[]|string
 ---@param options? InstallOptions
-function M.install(languages, options)
+M.install = a.sync(function(languages, options)
```
Comment (clason) [inline lua/nvim-treesitter/install.lua:569]: `a.sync` -> async, now I get it! :D

**Comment by lewis6991** [comment]
> What do you think of hard-coding the compiler + arguments per platform, and exposing a variable that allows providing a function that takes the necessary arguments (files, repo location, C++ scanner etc) and returns a custom compile invocation?
>
> (If we can somehow compile the norg parser without shenanigans, we can remove half of the infracstructure here...)

Emm, I haven't thought about that too much so I can't say either way. I'll keep it in mind and have a look.

Either way, the all the command pipeline stuff is gone now so this should be a lot easier to follow.

Messaging is still all over the place but maybe we accept that regression for now and follow up later?

**Comment by clason** [comment]
> Emm, I haven't thought about that too much so I can't say either way. I'll keep it in mind and have a look.

Yep, only tangentially related to this PR; I just wanted to make a note in case I forget.

> Messaging is still all over the place but maybe we accept that regression for now and follow up later?

Agreed. (Messaging was slightly broken already, what with the dreaded "press <enter>" prompt...)

I do think the way forward is to -- temporarily -- bundle a minimal(!) version of notifier.nvim for these progress messages.

File: scripts/write-lockfile.lua [hunk 3]
```diff
@@ -3,7 +3,7 @@
 vim.opt.runtimepath:append('.')
 
 -- Load previous lockfile
-local filename = require('nvim-treesitter.shell_cmds').get_package_path('lockfile.json')
+local filename = require('nvim-treesitter.install').get_package_path('lockfile.json')
```
Comment (clason) [inline scripts/write-lockfile.lua:6]: You didn't actually export that function...

**Review by clason — APPROVED** [review]
Squashy McSquashface?

**Comment by clason** [comment]
```
Error executing luv callback:
./lua/nvim-treesitter/async.lua:18: The coroutine failed with this message: ./lua/nvim-treesitter/install.lua:601: EEXIST: file already exists: /Users/clason/build/nvim-treesitter/runtime/queries/lua -> /Users/clason/.local/share/nvim/site/queries/lua
stack traceback:
        [C]: in function 'error'
        ./lua/nvim-treesitter/install.lua:601: in function <./lua/nvim-treesitter/install.lua:593>
stack traceback:
        [C]: in function 'error'
        ./lua/nvim-treesitter/async.lua:18: in function <./lua/nvim-treesitter/async.lua:12>
```

**Comment by lewis6991** [comment]
I've already fixed this in #4847

**Pull request merged**